src/process_definition_repository.ts

```diff
@@ -52,58 +52,61 @@ export class ProcessDefinitionRepository implements IProcessDefinitionRepository
 
   public async persistProcessDefinitions(name: string, xml: string, overwriteExisting: boolean = true): Promise<void> {
 
-    // Note:
-    // Unfortunately, sequelize doesn't have MIN/MAX operators for WHERE clauses.
-    // So in order to get the latest matching entry, we have to sort by the creation date and
-    // then cherry-pick the first entry.
-    const query: Sequelize.FindOptions<IProcessDefinitionAttributes> = {
-      limit: 1,
-      where: {
-        name: name,
-      },
-      order: [ [ 'createdAt', 'DESC' ]],
-    };
-
-    const processDefinitionHash: string = await this._createHashForProcessDefinition(xml);
-
-    const existingDefinitions: Array<ProcessDefinition> = await this.processDefinition.findAll(query);
-    const existingDefinition: ProcessDefinition = existingDefinitions.length > 0
-      ? existingDefinitions[0]
-      : undefined;
-
-    if (existingDefinition) {
-
-      if (!overwriteExisting) {
-        throw new ConflictError(`Process definition with the name '${name}' already exists!`);
-      }
-
-      // NOTE: bcrypt.compare() is broken. It ALWAYS returns true, no matter what you throw at it.
-      const hashesMatch: boolean = processDefinitionHash === existingDefinition.hash;
-      if (hashesMatch) {
-        // Hashes match: No changes were made.
-        // Just call "save" to update the "updatedAt" timestamp and move on.
-        await existingDefinition.save();
-
-        return;
-      }
-
-      // Hashes do not match: Changes were made.
-      // Create a new entry with the updated hash.
-      const createParams: any = {
-        name: name,
-        xml: xml,
-        hash: processDefinitionHash,
+    await this._sequelize.transaction(async(persistTransaction: Sequelize.Transaction): Promise<void> => {
+      // Note:
+      // Unfortunately, sequelize doesn't have MIN/MAX operators for WHERE clauses.
+      // So in order to get the latest matching entry, we have to sort by the creation date and
+      // then cherry-pick the first entry.
+      const findExistingDefinitionsQuery: Sequelize.FindOptions<IProcessDefinitionAttributes> = {
+        limit: 1,
+        where: {
+          name: name,
+        },
+        transaction: persistTransaction,
+        order: [ [ 'createdAt', 'DESC' ]],
       };
 
-      await this.processDefinition.create(createParams);
-    } else {
-
-      await this.processDefinition.create(<any> {
-        name: name,
-        xml: xml,
-        hash: processDefinitionHash,
-      });
-    }
+      const newProcessDefinitionHash: string = await this._createHashForProcessDefinition(xml);
+
+      const existingDefinitions: Array<ProcessDefinition> = await this.processDefinition.findAll(findExistingDefinitionsQuery);
+      const existingDefinition: ProcessDefinition = existingDefinitions.length > 0
+        ? existingDefinitions[0]
+        : undefined;
+
+      if (existingDefinition) {
+        if (!overwriteExisting) {
```

Comment:
And here.

Reply:
This already is a business rule.
It's just provided as a parameter.

Reply:
I think Vyperus meant that you should avoid an exclamation mark in an if-condition, to improve the readability.

For example:

```suggestion
		const doNotOverwriteExisting: boolean = !overwriteExisting;
        if (doNotOverwriteExisting) {
```

This way everyone can read the condition easily and doesn't run the risk of missing the exclamation mark.

Reply:
I think @Vyperus meant to avoid an exclamation mark (!) in if conditions.

For example:

```
		const doNotOverwriteExisting: boolean = !overwriteExisting;
        if (doNotOverwriteExisting) {
```

That way everyone can read the condition easily and doesn't run the risk of missing the exclamation mark.

Reply:
I get the picture.
But wrapping a business rule in a business rule, just to avoid some exclamation mark seems awfully redundant to me.

There is such a thing as overdoing it, imho.